src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h

```diff
@@ -0,0 +1,174 @@
+            deserializeItem(buf, value, arena);
+            values.insert(value);
+        }
+
+        narrowDown();
+    }
+
+    UInt64 threshold;
+    Storage values;
+};
+
+template <typename T>
+static void writeOneItem(WriteBuffer & buf, T item)
+{
+    if constexpr (std::is_same_v<T, StringRef>)
+        writeBinary(item, buf);
+    else
+        writeVarUInt(item, buf);
+}
+
+template <typename T>
+static void readOneItem(ReadBuffer & buf, Arena * arena, T item)
+{
+    if constexpr (std::is_same_v<T, StringRef>)
+        item = readStringBinaryInto(*arena, buf);
+    else
+        readVarUInt(item, buf);
+}
+
+template <typename T, bool is_weighted>
+struct AggregateFunctionGroupSortedArrayData
+{
+};
+
+template <typename T>
+struct AggregateFunctionGroupSortedArrayData<T, true> : public AggregateFunctionGroupSortedArrayDataBase<std::multimap<Int64, T>>
+{
+    using Base = AggregateFunctionGroupSortedArrayDataBase<std::multimap<Int64, T>>;
+    using Base::Base;
+
+    void add(T item, Int64 weight)
+    {
+        if (weight <= last || count < Base::threshold)
+        {
+            ++ count;
+            Base::values.insert({weight, item});
+            Base::narrowDown();
+            last = (--Base::values.end())->first;
+        }
+    }
+
+    void serializeItem(WriteBuffer & buf, typename Base::ValueType & value) const override
+    {
+        writeOneItem(buf, value.first);
+        writeOneItem(buf, value.second);
+    }
+
+    virtual void deserializeItem(ReadBuffer & buf, typename Base::ValueType & value, Arena * arena) const override
+    {
+        readOneItem(buf, arena, value.first);
+        readOneItem(buf, arena, value.second);
+    }
+
+    static T itemValue(typename Base::ValueType & value) { return value.second; }
+
+    Int64 last = std::numeric_limits<Int64>::max();
+    UInt64 count = 0;
+};
+
+
+template <typename T>
+class AggregateFunctionGroupSortedArrayDataMSet : public AggregateFunctionGroupSortedArrayDataBase<std::multiset<T>>
+{
+public:
+    using Base = AggregateFunctionGroupSortedArrayDataBase<std::multiset<T>>;
+    using Base::Base;
+
+    void serializeItem(WriteBuffer & buf, typename Base::ValueType & value) const override { writeOneItem(buf, value); }
+
+    void deserializeItem(ReadBuffer & buf, typename Base::ValueType & value, Arena * arena) const override
+    {
+        readOneItem(buf, arena, value);
+    }
+
+    static T itemValue(typename Base::ValueType & value) { return value; }
+};
+
+template <>
+struct AggregateFunctionGroupSortedArrayData<StringRef, false> : public AggregateFunctionGroupSortedArrayDataMSet<StringRef>
+{
+    using Base = AggregateFunctionGroupSortedArrayDataMSet<StringRef>;
+    using Base::Base;
+    void add(StringRef item)
+    {
+        Base::values.insert(item);
+        Base::narrowDown();
+    }
+};
+
+template <typename T>
+struct AggregateFunctionGroupSortedArrayData<T, false> : public AggregateFunctionGroupSortedArrayDataMSet<T>
```

Comment:
And for the rest of cases we optimize not playing with the map if the elements is not minor than the last value. Here the declaration for 'last' uses the same type 'T'.